cli/verifier/versions_match.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025 Edgeless Systems GmbH
+// SPDX-License-Identifier: BUSL-1.1
+
+package verifier
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/edgelesssys/contrast/internal/kuberesource"
+
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+)
+
+type identifier struct {
+	podRole string
+	image   string
+}
+
+var versionedResources = []identifier{
+	{image: "coordinator", podRole: "coordinator"},
+	{image: "contrast-initializer"},
+	{image: "contrast-service-mesh"},
+}
+
+func matchesResource(image, podRole string) bool {
+	for _, r := range versionedResources {
+		if r.image == image {
+			if podRole == "" || r.podRole == podRole {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// VersionsMatch verifies that the cli version matches the version of the used resources.
+type VersionsMatch struct {
+	Version string
+}
+
+// Verify verifies that the cli version matches the version of the used resources.
+func (v *VersionsMatch) Verify(toVerify any) error {
+	// ignore this verifier for preview and dev builds.
+	if strings.Contains(v.Version, "dev") || strings.Contains(v.Version, "pre") {
+		return nil
+	}
+
+	var findings error
+
+	kuberesource.MapPodSpecWithMeta(toVerify, func(
+		meta *applymetav1.ObjectMetaApplyConfiguration,
+		spec *applycorev1.PodSpecApplyConfiguration,
+	) (*applymetav1.ObjectMetaApplyConfiguration, *applycorev1.PodSpecApplyConfiguration) {
+		if spec.RuntimeClassName == nil || !strings.HasPrefix(*spec.RuntimeClassName, "contrast-cc") {
+			return meta, spec
+		}
+
+		for _, container := range spec.Containers {
+			if !matchesResource(*container.Name, meta.Annotations["contrast.edgeless.systems/pod-role"]) {
+				continue
+			}
+
+			image := *container.Image
+			if idx := strings.Index(image, "@"); idx != -1 {
+				image = image[:idx]
+			}
+			if idx := strings.Index(image, ":"); idx != -1 {
+				image = image[idx+1:]
+			}
```

Comment:
It would be safer to use a proper parser here (thinking of registry with host and port, for example) - I used https://pkg.go.dev/github.com/regclient/regclient/types/ref#Ref in the past, but maybe `go-containerregistry` has one, too?

Reply:
Hm, doesn't look like it. There's `Reference`, but that does not provided the wanted functionality. Went with regclient, that works nicely.